### ISSUE TITLE
Bug: [only-throw-error] only-throw-error.allow does not work when type is reexported

### ISSUE THREAD

**Issue by c-harding** [typescript-eslint/typescript-eslint#11817]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://eslint-online-playground.netlify.app/#eNp1UsGK2zAQ/ZVBFyeQ2LTQQ720BNJCb4XSW9WDkCeJs7IkJLmbxfjfO5Jsr7fp+mCJN2/ezLzRwLyTFd5EZxWWwbOatZ01LsAA30KwR9WiDrt0/+qccT/QW6M9wggnZzooDkKfeyVcJU3XGV1diFk8cL3ISBHkJeUuOe529ZHDdYOSchEkaQaIufWqbqJEsDxj2BRV71SxLW1rccM1rJQ3G8x5r3rcwqfP1EC4OPME1PGW6+0D27HgqdypPZdXbzRNPIwEoletDuUSIbyq4BD8Xl5QPsZOlpEaPLUaj4m6DJUFyIaIrhwIPkcmWni26KVrbdhnPBuBt0QmZdGrdC4V0qyzytSgLx1Gw5EcbH6S5DE2ic0ucof4A1BxM+KM321oyY56xgGscB7dPU4RZ64oQw3B9TjDY1J9OV2vcJ1WHO6GqoxWz/vk/B7jTooafr2UKTI26cVv1QOAUMo8UcKQPKuhsEI+0iTFDrToaNXF3a4pNJEo+t83CePvVb15FoAZzUj8b2kj9CImveWVxDBncVTOarp1piEjOEt5nDVo4za0bNFHwjQRZ/90k5PffSzflx84m2tmhT9f3hSZfL0mmDMlAvowFY+EHH8jeLef17zUBNcjG/8CzYtaGw==

### Repro Code

```TypeScript
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { catchError } from 'rxjs';

declare const http: HttpClient;

http.get('/url').pipe(
  catchError((e: HttpErrorResponse) => { throw e })
);
```

### ESLint Config

```javascript
// @ts-check

import { defineConfig } from 'eslint/config';
import tseslint from 'typescript-eslint';

export default defineConfig(
  tseslint.configs.recommendedTypeChecked,
  {
    languageOptions: {
      parserOptions: {
        project: true
      },
    },
    rules: {
      '@typescript-eslint/only-throw-error': [
        'error',
        {
          allow: [{ from: 'package', name: 'HttpErrorResponse', package: '@angular/common/http' }],
        },
      ],
    },
  },
);
```

### tsconfig

```jsonc
{}
```

### Expected Result

No error, regardless of the version of Angular installed.

### Actual Result

`@angular/common`: 19.2.0-19.2.4: no error

---

`@angular/common`: 19.2.5-19.2.x:
> Expected an error object to be thrown. [(@typescript-eslint/only-throw-error)](https://typescript-eslint.io/rules/only-throw-error) [5:48-49]

### Additional Info

`import { HttpClient, HttpErrorResponse, HttpParams } from '@angular/common/http';`

In newer versions of Angular, the package `@angular/common/http` contains the line:

```ts
export { /*...*/, HttpErrorResponse, *...*/ } from '../module.d-CnjH8Dlt.js';
```

This means that HttpErrorResponse is not defined directly in the file, but rather in another file. I can’t reference this other file, because the name is randomised.

**Comment by c-harding** [comment]
Update: it seems that it works using a prefix. All of the following configs work:

```js
          allow: [{ from: 'package', name: 'HttpErrorResponse', package: '@angular/common/module.d-' }],
          allow: [{ from: 'package', name: 'HttpErrorResponse', package: '@angular/common/module' }],
          allow: [{ from: 'package', name: 'HttpErrorResponse', package: '@angular/common/mod' }],
          allow: [{ from: 'package', name: 'HttpErrorResponse', package: '@angular/common/' }],
          allow: [{ from: 'package', name: 'HttpErrorResponse', package: '@angular/common' }],
          allow: [{ from: 'package', name: 'HttpErrorResponse', package: '@angular/co' }],
          allow: [{ from: 'package', name: 'HttpErrorResponse', package: '@angular/' }],
          allow: [{ from: 'package', name: 'HttpErrorResponse', package: '@angular' }],
          allow: [{ from: 'package', name: 'HttpErrorResponse', package: '@a' }],
          allow: [{ from: 'package', name: 'HttpErrorResponse', package: '@' }],
          allow: [{ from: 'package', name: 'HttpErrorResponse', package: '' }],
```

Is this intentional?

**Comment by JoshuaKGoldberg** [comment]
> This means that HttpErrorResponse is not defined directly in the file, but rather in another file

Fun! Yes, I think this is a bug in the core [TypeOrValueSpecifier](https://typescript-eslint.io/packages/type-utils/type-or-value-specifier) logic. It should support treating re-`export`s like this one as targets. Thanks for filing!

> works using a prefix ... Is this intentional?

No, that's odd. If you file a new issue for that we can take a look.

**Comment by dani1552** [comment]
Hi @JoshuaKGoldberg, I'd like to take a look at this and work on a fix.
Thanks @c-harding for the detailed report and reproduction!

**Comment by zaewc** [comment]
@c-harding @JoshuaKGoldberg 

From a first pass, this looks like it should be fixed in the shared `TypeOrValueSpecifier` package matching logic rather than only in `only-throw-error`

The failing case seems to be:

- `only-throw-error` calls `typeMatchesSomeSpecifier(type, allow, program)`.
- For `{ from: 'package', name: 'HttpErrorResponse', package: '@angular/common/http' }`, the package check currently follows the type's declaration files.
- In newer Angular versions, `HttpErrorResponse` is re-exported from `@angular/common/http`, but its actual declaration lives in a generated internal file such as `../module.d-*.js`.
- So the public import path matches the user's config, but the declaration-file-based package check no longer sees it as declared from `@angular/common/http`.

My intended direction is to add a regression test with a fake package that re-exports a type from an internal declaration file, then update the TypeOrValueSpecifier package matching logic to consider re-export/alias symbols when checking package specifiers.

I’ll keep the existing prefix/package-boundary behavior out of scope for this PR, since that looks related to the separate package matching issue.
